api/oc_push.c

```diff
@@ -0,0 +1,2555 @@
+  oc_recv_t *recv_obj = (oc_recv_t *)oc_memb_alloc(&recv_instance_memb);
+
+  if (!recv_obj) {
+    OC_PUSH_ERR("oc_memb_alloc() error!");
+    return result;
+  }
+
+  while (rep) {
+    switch (rep->type) {
+    case OC_REP_STRING:
+      if (!strcmp(oc_string(rep->name), "receiveruri")) {
+        oc_new_string(&recv_obj->receiveruri, oc_string(rep->value.string),
+                      oc_string_len(rep->value.string));
+        mandatory_property_check |= 0x1;
+      }
+      break;
+
+    case OC_REP_STRING_ARRAY:
+      if (!strcmp(oc_string(rep->name), "rts")) {
+        int len = oc_string_array_get_allocated_size(rep->value.array);
+        oc_new_string_array(&recv_obj->rts, len);
+
+        for (int i = 0; i < len; i++) {
+          oc_string_array_add_item(
+            recv_obj->rts, oc_string_array_get_item(rep->value.array, i));
+        }
+
+        mandatory_property_check |= 0x2;
+      }
+      break;
+
+    default:
+      OC_PUSH_ERR("something wrong, unexpected Property type: %d", rep->type);
+      break;
+    }
+    rep = rep->next;
+  }
+
+  if (mandatory_property_check != 0x3)
+    return result;
+  else
+    result = true;
+
+  oc_list_add(recvs_instance->receivers, recv_obj);
+
+  /* create app resource corresponding to receiver object */
+  OC_PUSH_DBG("new app resource for new receiver obj (\"%s\") is created...",
+              oc_string(recv_obj->receiveruri));
+  if (_create_pushd_rsc(recv_obj, recvs_instance->resource))
+    result = true;
+
+  return result;
+}
+
+/*
+ * @brief					replace existing receiver object array with new one
+ *
+ * @param recv_obj_list	receiver object array
+ * @param resource		Push Receiver resource
+ * @param rep				payload representation of new receiver object array
+ */
+static bool
+_replace_recv_obj_array(oc_recvs_t *recvs_instance, oc_rep_t *rep)
+{
+  int obj_arr_len;
+  oc_rep_t *rep_obj;
+  bool result = true;
+
+  /* remove existing receivers object array */
+  _purge_recv_obj_list(recvs_instance);
+
+  /* add new receivers object array */
+  while (rep) {
+    switch (rep->type) {
+    case OC_REP_OBJECT_ARRAY:
+      obj_arr_len = _get_obj_array_len(rep->value.object_array);
+      rep_obj = rep->value.object_array;
+
+      /* replace `receivers` obj array with new one */
+      for (int i = 0; i < obj_arr_len; i++, rep_obj = rep_obj->next) {
+        if (!_create_recv_obj(recvs_instance, rep_obj->value.object))
+          result = false;
+      } /* for */
+      break;
+
+    default:
+      OC_PUSH_ERR("something wrong, unexpected Property type: %d", rep->type);
+      break;
+    }
+
+    rep = rep->next;
+  } /* while */
+
+  return result;
+}
+
+/*
+ *
+ * @brief	POST callback for Push Receiver Resource
+ *
+ * @param request
+ * @param iface_mask
+ * @param user_data
+ */
+static void
+post_pushrecv(oc_request_t *request, oc_interface_mask_t iface_mask,
+              void *user_data)
+{
+  (void)iface_mask;
+  (void)user_data;
+
+  char *uri_param;
```

Comment:
`char *uri_param = NULL;`, because when `request->query` is `NULL` this value will not get initialized by oc_ri_get_query_value and it will have random, undefined value.

Reply:
agree, fixed